text/0000-unless-until.md

```diff
@@ -0,0 +1,586 @@
+    keywords `if` or `while` rather than the negative keywords `unless` or
+    `until`.
+
+    The negative words may lead casual readers to form improper assumptions
+    about control flow, inducing confusion or stutter when reading in more
+    depth.
+
+    If this RFC is accepted, the compiler MAY choose to add a lint warning that
+    is raised whenever an `unless !` or `unless { a } else { b }` construct is
+    encountered, and suggest changing them to an `if` or `if { b } else { a }`,
+    respectively.
+
+- Patterns *cannot* have interior bindings.
+
+    When an `unless let` or `until let` pattern matches, the branch governed by
+    it is **not** taken. As such, any bindings in the pattern would only be
+    accessible in blocks where they values to which they refer are **not**
+    alive.
+
+    As such, the following is invalid:
+
+    ```rust
+    unless let Err(e) = fallible() {
+        //  e is not in scope, because fallible() is not Err
+        //  the interior fields must be _
+    }
+    else {
+        //  e is accessible and in scope here, but it *should not be* in
+        //  scope, and NLL may later enforce this
+    }
+    ```
+
+    Patterns with interior data can be formed and inspected, but they cannot
+    bind:
+
+    ```rust
+    unless let Counter(x @ 1 ... 5) = expr() {
+        //  expr() might be a Counter(x > 5), OR any other variant!
+        //  Thus, the Counter interior data cannot be in scope
+    }
+    else {
+        //  Control jumps here when Counter(x @ 1 ... 5) matches, but if
+        //  you need access to the x binding, you should be using
+        //  `if let` because this is now the more interesting branch
+    }
+    ```
+
+    The guard clause can still be used, but without the `binding @` prefix:
+
+    ```rust
+    //  unnamed fields
+    unless let Counter(1 ... 5) = expr {}
+
+    //  named fields
+    struct Foo { x: i32 }
+    unless let Foo { x: 1 ... 5 } = expr {}
+    ```
+
+    This is compatible with existing Rust, where destructuring does not bind
+    unless an explicit `@` operator is used.
+
+    ```rust
+    let expr = Foo { x: 3 }
+    if let Foo { x: 1 ... 5 } = expr {
+        //  this branch enters, because expr.x is 3, but there is no
+        //  binding to x in scope
+    }
+    ```
+
+    If interior bindings are desired, this is a strong indication that your code
+    should be using `if let` or `while let` instead.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+- Why is this design the best in the space of possible designs?
+
+    Changing `let` bindings to have a negative operator such as `!=` is probably
+    way worse, since it seems Rust is explicitly trying to differentiate between
+    "these two concepts are logically equivalent" (`Eq` trait, `==` and `!=`
+    operators) and "this value is shaped like that pattern" (`let`, `match`).
+
+    Another concept is to introduce a *negative binding*, `!let`, which does not
```

Comment:
I personally would like to go all out and have both `if !let ...` and `if let ... && ...` work, making `let` in a conditional expression basically a `matches!` with binding. However, this does make "let-as-expression" more of a thing but it's still this weird kind of expression that's only an expression in certain contexts. So idk.